initial dev

kghatton opened the pull request.

Commit:
:tada: Initial code commit.

Largely based on the (v1) annotations-mapper service.

Commit:
:construction_worker: Add Circle config and Dockerfile.

Commit:
:memo: Updating readme.

Commit:
:memo: Updating readme.

Commit:
:memo: Updating readme.

Commit:
:white_check_mark: Readme link fixes, improve test coverage.

peteclark-ft commented on .circleci/config.yml:
```diff
@@ -0,0 +1,46 @@
+version: 2
```
This is just a suggestion, but I've added a circle workflow and a docker build job in the annotations-publisher (see [here](https://github.com/Financial-Times/annotations-publisher/pull/1/files#diff-1d37e48f9ceff6d8030570cd36286a61))

peteclark-ft commented on health/healthcheck.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (h *HealthCheck) checkKafkaConnectivity() (string, error) {
+	if err := h.consumer.ConnectivityCheck(); err != nil {
```
If the regex for the whitelist fails to compile, then this func will panic because the consumer will be `nil`

kghatton replied:
Not any more :)

peteclark-ft commented on main.go:
```diff
@@ -0,0 +1,161 @@
+		whitelist, err := regexp.Compile(*whitelistRegex)
+		if err != nil {
+			log.Error("Please specify a valid whitelist ", err)
+		} else {
```
See the above comment in the health code; if the whitelist regex fails to compile, we should either panic and shutdown, or provide a no-op unhealthy kafka consumer here

kghatton replied:
Yes, the startup and healthcheck code was copy/pasted from the annotations-mapper and I wasn't happy with it. Thank you for reminding me I needed to take another look at it :wink:.

peteclark-ft commented on service/messageHandler.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (mapper *AnnotationMapperService) HandleMessage(msg kafka.FTMessage) error {
+	tid := msg.Headers["X-Request-Id"]
```
Not 100% sure if this is true, but don't we strip the `X-` for kafka message headers? Or is it just the `X-Origin-System-Id` header that we do that for 🤔

kghatton replied:
This is copied from https://github.com/Financial-Times/annotations-mapper/blob/master/app.go#L118 - so if we use `X-Request-Id` there, I believe we should use it here too.

peteclark-ft commented on service/messageHandler_test.go:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	actual := service.HandleMessage(inbound)
+	assert.Error(t, actual, "error message")
```
Should this check the error message?

peteclark-ft replied:
I'd also leave the `"error message"` text out of the assertion, `assert.Error` is explanation enough of what is happening in that line

kghatton replied:
As the app itself isn't in control of the error message here, it's sensitive to changes in Go, so I'd rather not make an assertion about the text of the error message. Fair enough about the final parameter, though.

peteclark-ft commented on service/messageHandler_test.go:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	err := service.HandleMessage(inbound)
+	assert.EqualError(t, err, errmsg, "error message")
```
Similarly to above, `"error message"` doesn't give much value

peteclark-ft commented on service/messageHandler.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func (mapper *AnnotationMapperService) buildAnnotation(metadata PacMetadataAnnotation) annotation {
+	predicate := metadata.Predicate[strings.LastIndex(metadata.Predicate, "/")+1:]
```
Not sure how the other `annotations-mapper` works here, but should we validate that the predicate is one of our predicates? Otherwise, `http://www.ft.com/ontology/annotation/hasContributor` and `example.com/hasContributor` will produce valid predicates

Commit:
:ok_hand: Add Docker build and workflow to CircleCI config.

Commit:
:ok_hand: Remove superfluous test messages.

Commit:
:ok_hand: Validate predicates.

Commit:
:ok_hand: Handle a nil whitelist.

If there is no whitelist, then no messages will be mapped, and an error log will be written every time.

Commit:
:ok_hand: Proxy Kafka producer and consumer to allow unhealthy startup.

- Allow the app to startup and report unhealthiness if it can't connect to Kafka. Continue retrying to connect every minute until successful.
- Provide a real healthcheck for the consumer, by trying to connect as a distinct consumer on the same topic. The Kafka consumer library healthcheck is unimplemented.

galiri commented on service/conceptAnnotation.go:
```diff
@@ -0,0 +1,16 @@
+
+type annotation struct {
+	Thing      thing        `json:"thing"`
+}
```
why is this a Thing ? I'm assuming that isn't a neo Thing?

kghatton replied:
Renamed to `Concept`

galiri commented on service/conceptAnnotation.go:
```diff
@@ -0,0 +1,16 @@
+package service
+
+// ConceptAnnotations models the annotations as it will be written on the queue
+type ConceptAnnotations struct {
```
Naming - not sure what  ConceptAnnotation means. I understand what it is but the name is confusing as we always annotate with concepts. I think it's an Annotation (the binding) or maybe a MappedAnnotation.

kghatton replied:
Renamed to `MappedAnnotations`

galiri commented on service/messageHandler.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+func (mapper *AnnotationMapperService) HandleMessage(msg kafka.FTMessage) error {
+	tid := msg.Headers["X-Request-Id"]
```
Is it possible for the tid to be missing?

galiri commented on health/healthcheck.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+func (h *HealthCheck) checkKafkaConnectivity() (string, error) {
+	if err := h.consumer.ConnectivityCheck(); err != nil {
```
do we need to check the producer connection too?

peteclark-ft commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+FROM golang:1.8-alpine
```
Luca raised this on my PR for annotations-publisher - we could probably use the `golang:1.8.3-alpine` image, as this matches the golang version we use in circle.

izzyblues replied:
As I already commented in a Pete's PR, maybe it is good to use `golang:1.8.3-alpine` to make it consistent with circle...

peteclark-ft commented on health/healthcheck.go:
```diff
@@ -0,0 +1,81 @@
+		return gtgCheck(h.checkKafkaConnectivity)
+	}
+	producerCheck := func() gtg.Status {
+		return gtgCheck(h.checkKafkaConnectivity)
```
Not sure I follow this, is this not doing the same check twice?

peteclark-ft commented on kafka/consumerProxy.go:
```diff
@@ -0,0 +1,68 @@
+	return &ProxyConsumer{zookeeperConnectionString, consumerGroup, topics, config , nil}
+}
+
+func (c *ProxyConsumer) Connect() {
```
Ordinarily I'd say this probably should accept a `context.Context`, but since `StartListening` doesn't, then I'd leave it as it is!

peteclark-ft commented on kafka/consumerProxy.go:
```diff
@@ -0,0 +1,68 @@
+		return errors.New(errConsumerNotConnected)
+	}
+
+	// the Kafka consumer healthcheck is not reliable, so see if we can establish a new, independent connection
```
Why is it not reliable out of interest?

peteclark-ft replied:
Also, by checking a separate connection, we don't actually know if the component is connected to Kafka or not

kghatton replied:
@peteclark-ft See https://github.com/Financial-Times/kafka-client-go/blob/cc65f68255d42de893a6d72f31394ad031c54142/kafka/consumer.go#L87 for why the check is not reliable.

Commit:
:ok_hand: More intuitive naming.

Commit:
:ok_hand: Use "unknown" for logging if there is no X-Request-Id.

peteclark-ft commented on service/messageHandler.go:
```diff
@@ -0,0 +1,106 @@
+	tid := msg.Headers["X-Request-Id"]
+	requestLog := log.WithField("transaction_id", tid)
+	if mapper.whitelist == nil {
+		requestLog.Error("Skipping this message because the whitelist is invalid.")
```
Should the app fail or be unhealthy if the whitelist is invalid? If we do neither, we'll never process any messages, but the app will appear healthy

peteclark-ft replied:
Having said that, we would surely notice that the app wasn't processing any messages in pre-prod / lower envs, so it's probably fine like this.

Commit:
:ok_hand: Use updated base Docker image.

Commit:
:arrow_up: Upgrading to refactored Kafka client library (branch).

Commit:
:ok_hand: Renaming methods.

Commit:
:bug: Correcting predicate names.

Commit:
:arrow_up: Use release candidate of Kafka client library.

Commit:
:bug: Fixed application stop issue
